### PR TITLE
Add multihead attention

### PR DESCRIPTION
**Pull request by crcrpar**
This PR aims at supporting Multi-head attention including self attention in it in place of https://github.com/chainer/chainer/pull/7066 where I made a mistake while resolving conflicts.
I'm really sorry for this @himkt @hvy.

For the commit history, I made a copy of the branch https://github.com/crcrpar/chainer/pull/6.

Ref: 
- module: https://github.com/pytorch/pytorch/blob/76f847546b46890b9c028c7a98f36db8674bc0f5/torch/nn/modules/activation.py#L649-L783

- function: https://github.com/pytorch/pytorch/blob/76f847546b46890b9c028c7a98f36db8674bc0f5/torch/nn/functional.py#L3126-L3378

### CODE REVIEW

**Comment by himkt** [comment]
Thanks for revising the PR!

https://github.com/chainer/chainer/pull/7066#issuecomment-518667004
The reason is that this PR is related to the WIP PR https://github.com/cupy/cupy/pull/2215.

To supporting cuDNN MHA APIs, it could break compatibility with the current interface.
There are a few use-cases of cuDNN MHA and we need time to

**Comment by himkt** [comment]
Sorry, I accidentally closed this PR while writing a comment. :bow:
Reopened.

**Comment by himkt** [comment]
Thanks for revising the PR!

https://github.com/chainer/chainer/pull/7066#issuecomment-518667004
The reason is that this PR is related to the WIP PR https://github.com/cupy/cupy/pull/2215.

To supporting cuDNN MHA APIs, it could break compatibility with the current interface.
(this is why we think MHA should be experimental)
We need more time to resolve the issue https://github.com/cupy/cupy/issues/2202.
However, MHA plays an important role in current NLP community.
(this is why we think we should merge this PR)

**Comment by crcrpar** [comment]
OK, I got it.
But just I want say that I changed the design of
‘chainer.functions.multihead_attention’ while working on this PR to be
compatible with the cudnn imple as possible and this is why there are two
parameters for softmax scaler exist.

File: chainer/functions/connection/multihead_attention.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        - ``attn_output_weights`` is the weights of attention whose shape is
+          :math:`(B, L, S)` where :math:`B` is the batch size,
+          :math:`L` is the target sequence length,
+          and :math:`S` is the source sequence length. If ``return_weights`` is
+          ``False``, this return value is ``None``.
+
+    .. seealso:: :class:`~chainer.links.MultiHeadAttention`
+
+    """
+    chainer.utils.experimental('chainer.functions.multihead_attention')
```
Comment (himkt) [inline chainer/functions/connection/multihead_attention.py]: Please add a statement `import chainer` on the top of this module.

File: chainer/links/__init__.py [hunk 2]
```diff
@@ -24,6 +24,7 @@
 from chainer.links.connection.linear import Linear  # NOQA
 from chainer.links.connection.local_convolution_2d import LocalConvolution2D  # NOQA
 from chainer.links.connection.mlp_convolution_2d import MLPConvolution2D  # NOQA
+from chainer.links.connection.multihead_attention import MultiHead Attention  # NOQA
```
Comment (himkt) [inline chainer/links/__init__.py]: Should be `import MultiHeadAttention`.

**Comment by himkt** [comment]
(I'm so sorry I failed submitting response comments. :bow:)

> I changed the design of `chainer.functions.multihead_attention` while working on this PR to be compatible with the cudnn imple as possible
Sounds amazing! Thanks for letting me know. I remember this.

File: chainer/links/__init__.py (hunk 2 above)
Comment (crcrpar) [reply chainer/links/__init__.py]: Thanks.

File: chainer/functions/connection/multihead_attention.py (hunk 1 above)
Comment (crcrpar) [reply chainer/functions/connection/multihead_attention.py]: I think you meant to add this comment to chainer/links/connection/multihead_attention.py.

**Comment by crcrpar** [comment]
Only mac OS python2 test failed, meaning this PR itself is ready for review I think.

**Comment by himkt** [comment]
@hvy Please review it. :bow:

File: chainer/functions/connection/multihead_attention.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    return_weights=True       # type: bool
+):
+    # type: (...) -> tp.Tuple[variable.Variable, variable.Variable]  # NOQA
+    """Multi-head Attention forward function.
+
+    Args:
+        query (:class:`~chainer.Variable` or :ref:`ndarray`):
+            A batch of query vectors whose shape is :math:`(L, B, E)` where
+            :math:`L` is the target sequence length, :math:`B` is the
+            batch size, and :math:`E` is the embedding size.
```
Comment (hvy) [inline chainer/functions/connection/multihead_attention.py]: Could you explain why the batch dimension isn't placed first?

File: chainer/functions/connection/multihead_attention.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        attn_output = dropout.dropout(attn_output, post_dropout)
+
+    if return_weights:
+        attn_output_weights = reshape.reshape(
+            attn_output_weights,
+            (batch_size, n_head, target_length, source_length)
+        )
+        attn_output_weights = average.average(attn_output_weights, axis=1)
+    else:
+        attn_output_weights = None
```
Comment (hvy) [inline chainer/functions/connection/multihead_attention.py]: Could you motivate why it's better to return `None` instead of not retuning this at all? Just curious because the latter is more commonly in existing functions such as dropout and Gaussian sampling.

File: chainer/links/connection/multihead_attention.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+                self.bias_k, self.bias_v = None, None
+
+    def forward(
+        self,
+        query,                     # type: InputType
+        key=None,                  # type: tp.Optional[InputType]
+        value=None,                # type: tp.Optional[InputType]
+        key_padding_mask=None,     # type: tp.Optional[InputType]
+        attention_mask=None,       # type: tp.Optional[InputType]
+        add_zero_attention=False,  # type: bool
```
Comment (hvy) [inline chainer/links/connection/multihead_attention.py]: This argument is not documented.

File: chainer/functions/connection/multihead_attention.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+        expected_embedding_size (int): Total number of units of the model.
+        n_head (int): The number of parallel attention heads.
+        proj_in_W (:obj:`tuple`, :class:`~chainer.Variable` or :ref:`ndarray`):
+            Weight(s) to project `query`, `key`, and `value` vectors.
+            If the input sizes of `query`, `key`, and `value` are different,
+            this should be the tuple of three weights, otherwise, one weight.
+        proj_in_b (:class:`~chainer.Variable` or :ref:`ndarray`):
+            Bias added to projected `query`, `key`, and `value` vectors.
+        add_zero_attn (bool): If ``True``, add a new batch of zeros to
+            the key and value sequences at axis=1.
```
Comment (hvy) [inline chainer/functions/connection/multihead_attention.py]: How about elaborating a bit on this description, when it should be used?

**Comment by hvy** [comment]
I left a couple of minors comments on details. I'm not confident enough to comment on the public interface so I asked @soskek for some feedback (from a researcher/user perspective).

File: chainer/functions/connection/multihead_attention.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+    proj_out_W,               # type: variable.Variable
+    proj_out_b,               # type: variable.Variable
+    add_zero_attn=False,      # type: bool
+    attention_dropout=0,      # type: float
+    post_dropout=0,           # type: float
+    key_padding_mask=None,    # type: tp.Optional[InputType]
+    attn_mask=None,           # type: tp.Optional[InputType]
+    dot_product_scaler=None,  # type: tp.Optional[float]
+    softmax_scaler=1.0,       # type: float
+    return_weights=True       # type: bool
```
Comment (soskek) [inline chainer/functions/connection/multihead_attention.py]: Are you planning to implement `static_k` and `static_v`, which are implemented in Pytorch?
While their name is a little confusing, they are the values computed in here:
https://github.com/chainer/chainer/blob/723fdd97b47e8848569dfa59f0ff655eaad1c5a1/chainer/functions/connection/multihead_attention.py#L203-L206

If we use the same k and v many times, we can reuse the values without recomputing. The static_k and _v are prepared for the purpose. For example, a typical transformer for machine translation encodes a source sentence via self-attention and outputs sequential vectors, and in its decoder, (in addition to self-attention) it performs attention over the vectors in each layer. In this case, the decoder attends the same set of vectors (for k and v) many times.

In my opinion, these features are just additional. And, we can omit it in this first PR.

(cf. In pytorch and its original source, fairseq, the arguments are used in these lines.)
https://github.com/pytorch/pytorch/blob/883628cb5c31c89323cef52e96310c92abecfcbe/torch/nn/functional.py#L3318-L3326
https://github.com/pytorch/fairseq/blob/3e3fe72299980f53262880e24e372ed7d785093c/fairseq/modules/multihead_attention.py#L180

File: chainer/links/connection/multihead_attention.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+                (time, batch_size, value_in_size).
+            key_padding_mask (:class:`~chainer.Variable` or :ref:`ndarray`):
+                If not ``None``, mask the memory slots.
+                The shape is (batch_size, source_length).
+                Each value is 0 (``False``) or 1 (``True``) where 1 means that
+                the memory slot will not be used.
+            attention_mask (:class:`~chainer.Variable` or :ref:`ndarray`):
+                Mask help attention ignores certain positions.
+                The shape is :math:`(L, L)` where :math:`L` is
+                the target sequence length.
```
Comment (soskek) [inline chainer/links/connection/multihead_attention.py]: Ditto. Adding shape and explanation as an additive mask.

File: chainer/functions/connection/multihead_attention.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+        post_dropout (float): Dropout ratio at the output.
+        proj_out_W (:class:`~chainer.Variable` or :ref:`ndarray`):
+            Weight to project attention.
+        proj_out_b (:class:`~chainer.Variable` or :ref:`ndarray`):
+            Bias for projected attention.
+        key_padding_mask (:class:`~chainer.Variable` or :ref:`ndarray`):
+            If not ``None``, specified padding elements in the key
+            will be ignored by the attention.
+            The shape is :math:`(B, S)` where :math:`B` is the batch size,
+            and :math:`S` is the source sequence length.
```
[the remark this answers is not in the document]
Comment (soskek) [reply chainer/functions/connection/multihead_attention.py]: Can we make a description in `chainer/links/connection/multihead_attention.py` consistent with this part?

File: chainer/links/connection/multihead_attention.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+                The key of the memory vectors with the shape of
+                (time, batch_size, key_in_size).
+            value (:class:`~chainer.Variable` or :ref:`ndarray`):
+                The value of the memory with the shape of
+                (time, batch_size, value_in_size).
+            key_padding_mask (:class:`~chainer.Variable` or :ref:`ndarray`):
+                If not ``None``, mask the memory slots.
+                The shape is (batch_size, source_length).
+                Each value is 0 (``False``) or 1 (``True``) where 1 means that
+                the memory slot will not be used.
```
Comment (soskek) [inline chainer/links/connection/multihead_attention.py]: Please add the shape.

File: chainer/functions/connection/multihead_attention.py (hunk 4 above)
Comment (crcrpar) [reply chainer/functions/connection/multihead_attention.py]: I don't think it's better or not, but I wanted to make the return type consistent (`tuple`).

File: chainer/functions/connection/multihead_attention.py (hunk 7 above)
Comment (crcrpar) [reply chainer/functions/connection/multihead_attention.py]: Thank you for your clarification of `static_k` and `static_v`.

IIRC, I tried to implement those `static` first, but I gave it up to avoid making this PR too complicated. Also, I thought it reasonable to create a follow-up PR of improved unit tests and `static_k` and `static_v`.

What do you think?

File: chainer/functions/connection/multihead_attention.py (hunk 3 above)
Comment (crcrpar) [reply chainer/functions/connection/multihead_attention.py]: @hvy Does the above explanation make sense to you?

**Comment by crcrpar** [comment]
I'll resume this.

File: chainer/functions/connection/multihead_attention.py (hunk 7 above)
Comment (soskek) [reply chainer/functions/connection/multihead_attention.py]: I agree with you. `static_` stuff would make this PR complicated. It seems good to skip now.

File: chainer/functions/connection/multihead_attention.py [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+            batch size, and :math:`E` is the embedding size.
+        key (:class:`~chainer.Variable` or :ref:`ndarray`)
+            A batch of key vectors whose shape is :math:`(S, B, E)` where
+            :math:`S` is the source sequence length, :math:`B` is the
+            batch size, and :math:`E` is the embedding size.
+        value (:class:`~chainer.Variable` or :ref:`ndarray`)
+            A batch of value vectors whose shape is :math:`(S, B, E)` where
+            :math:`S` is the source sequence length, :math:`B` is the
+            batch size, and :math:`E` is the embedding size.
+        expected_embedding_size (int): Total number of units of the model.
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply chainer/functions/connection/multihead_attention.py]: It seems you are right. Thanks.

File: chainer/functions/connection/multihead_attention.py [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+            Bias for projected attention.
+        key_padding_mask (:class:`~chainer.Variable` or :ref:`ndarray`):
+            If not ``None``, specified padding elements in the key
+            will be ignored by the attention.
+            The shape is :math:`(B, S)` where :math:`B` is the batch size,
+            and :math:`S` is the source sequence length.
+        attn_mask (:class:`~chainer.Variable` or :ref:`ndarray`):
+            Mask help attention ignores certain positions.
+            The shape is :math:`(L, L)` where :math:`L` is
+            the target sequence length.
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply chainer/functions/connection/multihead_attention.py]: > This shape should be `(L, S)`?

Thanks.



> And, add an explanation that this value is "added" into attention values before softmax. Maybe, typical users expect the format of a mask for ignorance as multiplication. Adding a typical usage (e.g. -inf means setting ignoring the position) could be nice.

Yes, yes, yes!
I got confused why masking was done by addition instead of multiplication.

**Comment by crcrpar** [comment]
Future follow-up PRs should access

- implementing `static_k` and `static_v`
- improving unit tests

File: chainer/functions/connection/multihead_attention.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (crcrpar) [reply chainer/functions/connection/multihead_attention.py]: Use the notation of `F.multi_head_attention` in `L.MultiHeadAttention` for fewer characters hoping it'll make the documentation more readable.

File: chainer/functions/connection/multihead_attention.py [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+
+    def _in_proj_v(value):
+        return _in_proj(value, start=2 * embedding_size, weight_idx=2)
+
+    if embedding_size % n_head != 0:
+        raise ValueError(
+            '`embedding_size` ({}) need to be '.format(embedding_size) +
+            'divisible by `n_head` ({})'.format(embedding_size, n_head))
+    if (bias_k is None) != (bias_v is None):
+        raise ValueError
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply chainer/functions/connection/multihead_attention.py]: Sorry for being slow.

I agree that I have better adding a certain message. However, I'm not sure what kind of message is kind enough here.
Therefore, I just added a message saying that one bias is not `None` and the other is not `None`.

File: chainer/functions/connection/multihead_attention.py [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+            k, v = _in_proj_kv(key)
+    else:
+        q = _in_proj_q(query)
+        k = _in_proj_k(key)
+        v = _in_proj_v(value)
+    q *= dot_product_scaler
+
+    if bias_k is not None:
+        k = concat.concat((k, repeat.repeat(bias_k, batch_size, axis=1)))
+        v = concat.concat((v, repeat.repeat(bias_v, batch_size, axis=1)))
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply chainer/functions/connection/multihead_attention.py]: I'm referencing https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L3294-L3297.

Will recheck this, including the difference between the behavior of `chainer.functions.repeat` and that of `torch.Tensor.repeat`.

[the remark this answers is not in the document]
Comment (crcrpar) [reply chainer/functions/connection/multihead_attention.py]: > These concat-s should be done in axis=0 rather than =1 (default)?

Looks it should. ([`chainer.functions.concat`](https://docs.chainer.org/en/stable/reference/generated/chainer.functions.concat.html) -> axis=1, [`torch.cat`](https://pytorch.org/docs/stable/torch.html#torch.cat) -> axis=0.)

**Comment by kmaehashi** [comment]
As Chainer is now under the maintenance phase, unfortunately we decided not to merge this feature. Thanks again for your contribution.
